pkg: Semtech LoRaMAC

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR imports the Semtech LoRaMAC reference implementation as a package and proposes a simple API for interacting with a LoRaWAN network.

What has been tested so far:
* OTAA and ABP activation
* Emission and reception of confirmed and unconfirmed messages
* Message reception sent from the LoRaWAN backend
* Adaptive datarate

This PR also provides a test application for setting/getting some LoRaMAC parameters, joining to a network and sending/receiving data to/from the network.

This PR has been tested with SX1272 and SX1276 devices, using Nucleo 64 boards and the b-l072z-lrwan1 board.

Of course there's room for improvement:
* better management for errors/problems
* adaption to netdev is missing and it's still unclear to me how to achieve this
* add other LoRaWAN functionnalities: more parameters (RX2 delays, etc), link check and LoRaWAN compliance tests

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Issues/PRs references

refs #7331 

This PR is based on ~#7592~ and ~#8261~ 

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

PeterKietzmann commented:
Is this an alternative to #6645?

aabadie commented:
I also forgot to mention that this PR is an adaption of an initial work from @jia200x. So most of the credits go to him.

miri64 commented:
> I also forgot to mention that this PR is an adaption of an initial work from @jia200x. So most of the credits go to him.

Why doesn't the commit authorship then? ;-)

aabadie commented:
> Why doesn't the commit authorship then? ;-)

Lot's of refactoring adaption so it was not easily possible. But I kept the credits in the files headers.

miri64 commented:
As long as @jia200x is fine with that, I don't mind.

aabadie commented:
@kaspar030 @smlng any idea why Murdock is failing here ? because locally it builds fine for me. I'm wondering if some git caching could be involved.

dylad commented:
Is there a way to test  this without setting up the whole TTN stuff ? Can we test it with Semtech tools on a gateway ?

aabadie commented:
> Is there a way to test this without setting up the whole TTN stuff ?

There are other LoRaWAN providers (Actility, LoRiot, etc) and I think you can also setup your own LoRaWAN server (see [here](https://www.loraserver.io/)).

> Can we test it with Semtech tools on a gateway ?

Maybe yes but I have to admit that I don't know how. Looking at loraserver.io, they also provide a [LoRa gateway bridge application](https://docs.loraserver.io/lora-gateway-bridge/overview/) that can run on the gateway itself. That looks interesting.

aabadie commented:
> any idea why Murdock is failing here ?

I found the problem, it was in the package Makefile. Now there are only platform related issues remaining.

aabadie commented:
Ci is now fixed ! I had to blacklist the MSP430 based boards because of the following issue:
```
semtech_loramac_crypto.a(aes.o):(.debug_loc+0x596): relocation truncated to fit: R_MSP430_16_BYTE against `no symbol'
collect2: ld a retourné 1 code d'état d'exécution
```

Not sure how and if it can be fixed.

miri64 commented:
Needs rebase

aabadie commented:
rebased

miri64 commented on pkg/semtech-loramac/include/semtech-loramac/timer.h:
```diff
@@ -0,0 +1,130 @@
+    xtimer_t dev;       /**< xtimer instance attached to this LoRaMAC timer */
+    msg_t msg;          /**< message attacher to this LoRaMAC timer */
+    void (*cb)(void);   /**< callback to call when timer timeout */
+} TimerEvent_t;
```
Coding convention (`slug_case` instead of CamelCase; also the opening `{` is of)?

aabadie replied:
This is normal since this file is included by the internal LoRaMAC patched code.

aabadie replied:
I can fix the opening `{` though

miri64 replied:
Ah, alright then!

miri64 commented on pkg/semtech-loramac/patches/0002-adapt-for-RIOT.patch:
```diff
@@ -0,0 +1,281 @@
+-    TimerSetValue( &MacStateCheckTimer, 1 );
+-    TimerStart( &MacStateCheckTimer );
++    // TimerSetValue( &MacStateCheckTimer, 1 );
++    // TimerStart( &MacStateCheckTimer );
```
Rather remove instead of comment out, otherwise porting to later versions might more easily provoke conflicts in these lines.

aabadie replied:
done

miri64 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,690 @@
+    _main_pid = thread_getpid();
+    _mac_pid = thread_create(stack, sizeof(stack), THREAD_PRIORITY_MAIN - 1,
+                             THREAD_CREATE_STACKTEST, _event_loop, NULL,
+                             "recv_thread");
```
IMHO we should do aim for one of the following:

1. have LoRAMAC be a netdev layer
2. have LoRAMAC be a `gnrc_netif` thread

From the current code base the latter would be easier to port to and my preferred method (since layered netdev is only WIP at the moment). In any case LoRAMAC should not be a completely separate thread from `gnrc_netif`.

miri64 commented:
In the current state this needs to be somewhat restructured in my opinion (that is why I set it to WIP). Having a dedicated thread just for LoRAMAC below the interface layer (at least that seems like what you are planning to do) is overkill.

aabadie commented:
Moving the following comment in the main discussion:

> IMHO we should do aim for one of the following:
>
>  * have LoRAMAC be a netdev layer
>  * have LoRAMAC be a gnrc_netif thread
>
>From the current code base the latter would be easier to port to and my preferred method (since layered netdev is only WIP at the moment). In any case LoRAMAC should not be a completely separate thread from gnrc_netif.

I can try to adapt to gnrc_netif but I need some advice for this.

miri64 commented:
> I can try to adapt to gnrc_netif but I need some advice for this.

Don't be afraid to ask :-)

miri64 commented:
> You can consider my previous comment a question then :)

Which one?

aabadie commented:
this one:
> I can try to adapt to gnrc_netif but I need some advice for this.

miri64 commented:
> this one:
>
>> I can try to adapt to gnrc_netif but I need some advice for this.

Still not sure how to help you then...

miri64 commented:
Start by providing the functions required by the `gnrc_netif_ops_t` struct first I would say, then create a `gnrc_netif_semtec_loramac_create` function to initialize it and work your way from there.

jia200x commented:
Great!

I'm very busy atm since I leave my country in some days, but as soon as I
find some time I will review it.

El 19 dic. 2017 6:40, "Martine Lenders" <notifications@github.com> escribió:

> Start by providing the functions required by the gnrc_netif_ops_t struct
> first I would say, then create a gnrc_netif_semtec_loramac_create
> function to initialize it and work your way from there.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/8264#issuecomment-352690488>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABM8SPshLLjmv9KZ6cXOznZXkiedHxQkks5tB4SAgaJpZM4RB01H>
> .
>

miri64 commented:
> Thanks. Should I put code in the pkg directory ?

Yes. I think you just have to touch `pkg/semtech-loramac/contrib/semtech_loramac.c` for this.

jia200x commented:
> I also forgot to mention that this PR is an adaption of an initial work from @jia200x. So most of the credits go to him.

Also, @fjmolinas coworked with the initial part, so credits go to him as well :)

aabadie commented:
> Start by providing the functions required by the gnrc_netif_ops_t struct first I would say, then create a gnrc_netif_semtec_loramac_create function to initialize it and work your way from there.

@miri64, I tested the netif integration (not pushed) but I have a firmware size problem when I activated it. With the actual status of this PR, the RAM/ROM is about 7kB/55kB with the netif thread (I have to use gnrc_netdev_default to build), it's 13kB/73kB. This is way too big.

Is there a way to limit what is picked from netif (since I don't want - yet - all pktbuf, netreg, netapi, etc) ?

aabadie commented:
> Is there a way to limit what is picked up from netif (since I don't want - yet - all pktbuf, netreg, netapi, etc) ?

@miri64 ?

miri64 commented:
> Is there a way to limit what is picked from netif (since I don't want - yet - all pktbuf, netreg, netapi, etc) ?

It doesn't make much sense (application-wise) to compile `gnrc_netif` without the other GNRC components, so no there is no default way to do that. What you can try is to remove the 

```Makefile
ifneq (,$(filter gnrc_%,$(filter-out gnrc_netapi gnrc_netreg gnrc_netif% gnrc_pkt%,$(USEMODULE))))
  USEMODULE += gnrc
endif
```

from the Makefile.dep and just to include `gnrc_netif` (and maybe `gnrc_netif_hdr`), but I can't promise that it will work (or even compile)

aabadie commented:
> just to include gnrc_netif (and maybe gnrc_netif_hdr), but I can't promise that it will work (or even compile)

I tried that and it doesn't work without patching the `_pass_packet` function in `gnrc_netif.c` and the gnrc_netif function used in the shell module (because the test application uses the shell).

In the end, I could build a similar sized firmware but it doesn't work because of the way the MAC is initialized.

By the way, I just pushed a couple of improvements of this PR:
* removed the forced number of join retries in EU868: it was 48 minimum, which is way too big and takes ages if the join parameters are invalid. The default is now 3
* removed the loramac params (for euis and keys) and getter/setter functions for setting them

kaspar030 commented on pkg/semtech-loramac/Makefile.include:
```diff
@@ -0,0 +1,9 @@
+            -I$(PKGDIRBASE)/semtech-loramac/src/system/crypto \
+            -I$(PKGDIRBASE)/semtech-loramac/src
+
+CFLAGS += -Wno-missing-field-initializers -Wno-unused-parameter -Wno-sign-compare
```
Are these cflags needed for the headers? If not, move this to Makefile

kaspar030 commented on pkg/semtech-loramac/Makefile.include:
```diff
@@ -0,0 +1,9 @@
+INCLUDES += -I$(RIOTBASE)/pkg/semtech-loramac/include \
```
Are all of these needed for *using* loramac? Everything internal to the module should go into Makefile.

aabadie replied:
Yes, it seems to be the case. I tried different things but none worked (I'm not a build system expert though)

kaspar030 replied:
This has nothing to do with the build system. If added this way, all these includes will be applied *globally* (to all modules). Please confirm that this is necessary, or maybe can be shrunk a little.

kaspar030 commented on pkg/semtech-loramac/Makefile.dep:
```diff
@@ -0,0 +1,6 @@
+USEMODULE += random
```
This file has the same semantics as ```Makefile.dep``` in the root folder. So if someone does ```USEPKG += semtech-loramac```, all these modules should get selected?

aabadie replied:
yes, do you suggest that I should add an entry in the root Makefile.dep and remove this file ?

aabadie replied:
Looking at this file, the actual way of including dependencies seems correct

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac_arch.c:
```diff
@@ -0,0 +1,67 @@
+    return random_uint32_range(min, max + 1);
+}
+
+void memcpy1(uint8_t *dst, const uint8_t *src, uint16_t size)
```
Why is this needed?

kaspar030 replied:
Where are memcpy1 and memset1 declared? Maybe that declaration can be patched into calling the proper functions.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac_arch.c:
```diff
@@ -0,0 +1,67 @@
+    }
+}
+
+void memset1(uint8_t *dst, uint8_t value, uint16_t size)
```
and this?

jia200x replied:
Those are STM32 specific code that's being called from LoRaMAC. we would need to patch every memset1 line to memset in order to remove this function.

kaspar030 replied:
or maybe add ```#define memset1 memset``` at a strategic position?

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac_arch.c:
```diff
@@ -0,0 +1,67 @@
+ * @}
+ */
+
+#include "semtech-loramac/board.h"
```
why is this needed?

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac_radio.c:
```diff
@@ -0,0 +1,246 @@
+
+#include "semtech-loramac/board.h"
+
+#include "radio/radio.h"
```
remove extra newline, here and below

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac_radio.c:
```diff
@@ -0,0 +1,246 @@
+
+extern sx127x_t sx127x;
+
+#define LORAMAC_RX_WINDOW_DURATION  (600 * US_PER_MS)
```
600LU

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac_timer.c:
```diff
@@ -0,0 +1,86 @@
+#include "xtimer.h"
+#include "thread.h"
+
+extern kernel_pid_t _mac_pid;
```
is there only one _mac_pid? Variable might need to be namespaced.

kaspar030 commented on pkg/semtech-loramac/include/semtech_loramac.h:
```diff
@@ -0,0 +1,277 @@
+ *
+ * @param[in] eui           The device EUI
+ */
+void semtech_loramac_set_deveui(uint8_t *eui);
```
here and below for all setters: can the arguments be const?

kaspar030 commented on pkg/semtech-loramac/patches/0001-add-RIOT-makefiles.patch:
```diff
@@ -0,0 +1,57 @@
+From d45cbc53687d090a69cf669e8e8a9fb5460db943 Mon Sep 17 00:00:00 2001
```
Please don't use a patch here. Put the Makefiles into the package directory and just copy them over in the compilation step (see how e.g., heatshrink does it). Way easier to maintain.

kaspar030 commented on tests/pkg_semtech-loramac/main.c:
```diff
@@ -0,0 +1,362 @@
+#include "sx127x_netdev.h"
+
+sx127x_t sx127x;
+static uint8_t rx_buf[242] = { 0 };
```
no need to initialize to 0, C guarantees that for us implicitly.

kaspar030 commented on tests/pkg_semtech-loramac/main.c:
```diff
@@ -0,0 +1,362 @@
+    };
+}
+
+void _bytes_to_hex(const uint8_t *byte_array, char *hex, uint8_t max_len)
```
use fmt?

aabadie replied:
I thought of it but in fmt there's only the logic for one byte or for an array of bytes but reversed. This is not what I'm doing here. Maybe that would deserve a PR for fmt ?

kaspar030 replied:
You could use just the one-byte version in a loop. ;)

aabadie replied:
@kaspar030, updated to use the new `fmt` functions. Works like a charm!

miri64 replied:
> Works like a charm!

That's the benefit of proper testing ;-)

jia200x commented:
> In the current state this needs to be somewhat restructured in my opinion (that is why I set it to WIP). Having a dedicated thread just for LoRAMAC below the interface layer (at least that seems like what you are planning to do) is overkill.

@miri64 the _event_loop function handles part of the LoRaMAC logic. Since currently almost all LoRaWAN applications interact directly with the MAC layer (that's how LoRaWAN works), I would just leave it as it until we continue with more upper layers in the future (e.g IPv6 on top of LoRaWAN). 

For that, we would need to remove the _event_loop thread, add the proper handlers for LoRaMAC timers, reception windows, among others; and also we might redefine the interface between RIOT and Semtech code.

miri64 commented:
> @miri64 the _event_loop function handles part of the LoRaMAC logic. Since currently almost all LoRaWAN applications interact directly with the MAC layer (that's how LoRaWAN works), I would just leave it as it until we continue with more upper layers in the future (e.g IPv6 on top of LoRaWAN).

Why not safe us the work now and implement LoRAMAC directly as a `gnrc_netif` enabled thread? If you worry about merging LoRAMAC too much into GNRC, you can also provide a module that is called by that `gnrc_netif` thread without introducing an extra thread.

jia200x commented:
`Why not safe us the work now and implement LoRAMAC directly as a gnrc_netif enabled thread? If you worry about merging LoRAMAC too much into GNRC, you can also provide a module that is called by that gnrc_netif thread without introducing an extra thread.`

I think that's what @aabadie is already doing in [his branch]. The proposition was just because GNRC components are not being used atm in this PR, so there won't be an extra thread in the same elf file. When we add the upper parts (netapi, 6lowpan), then I think is mandatory to move the _event_loop logic to gnrc_netif thread.

What do you think?

[his branch]:https://github.com/aabadie/RIOT/tree/pkg_semtech-loramac-netif

aabadie commented:
> What do you think?

I agree :)

@kYc0o, I also reworked the test application as you suggested offline. Now all is done in a single `loramac` commands with subcommands (`get`, `set`, `join`, `tx`). I also tried to improve the usage functions to guide the user.

aabadie commented:
@miri64, as @jia200x mentioned above the actual state of this PR is out of gnrc so having a separate thread in the package is safe until we start working on the netif integration (like I tried [here](https://github.com/aabadie/RIOT/tree/pkg_semtech-loramac-netif)). I think reusing the netif thread would be done in a follow-up PRs.

So, for me, this PR is not WIP anymore. I tested most of LoRaWAN configurations with success (at least for EU868). Are you ok to remove the WIP label and try to have it merged before wednesday ?

smlng commented:
@miri64 and. @kaspar030 please check if you comments are sufficiently addressed - meaning would: be nice to have the LoRa _feature_ for release 2018.01. And as external package, it looks like this doesn't impacts any internal part of RIOT, right?

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+#endif
+
+static char stack[LORAMAC_STACKSIZE];
+kernel_pid_t _semtech_loramac_pid;
```
The next ten lines: can any of the variables be static? Or namespaced?

aabadie replied:
Maybe the keys could be made static in a header file (I'm not sure) since they need to be reachable from different c files. For the moment, `extern` is used.
For `radio_events`, I don't think this is easily doable, because that would require including LoRamac.h from a header file, and this should avoided (otherwise there will be name clashes).

aabadie replied:
Is it fine, if I go for namespaces ?

kaspar030 replied:
Yes please. All global variables have to be namespaced. All file scope (static) variables *should* start with underscore followed by non-capital letter.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+static bool _semtech_loramac_send(uint8_t cnf, uint8_t port, uint8_t dr,
+                                  uint8_t *payload, uint8_t len)
+{
+    DEBUG("[semtech-loramac] send frame %s\n", (char*)payload);
```
```(char *)payload```. I suggest running uncrustify.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+        return;
+    }
+
+#if ENABLE_DEBUG
```
```if (ENABLE_DEBUG)``` (no macro)

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+        memcpy(rx_payload, indication->Buffer, indication->BufferSize + 1);
+        rx_port_internal = indication->Port;
+        DEBUG("[semtech-loramac] MCPS indication, RX data: %s, Port: %d\n",
+              (char*)rx_payload, rx_port);
```
```char *```

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+static void _join(void *arg)
+{
+    (void) arg;
+    uint8_t join_type = *(uint8_t*)arg;
```
```*(uint8_t *)arg```

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+    msg.content.ptr = dev;
+
+    switch (event) {
+
```
remove empty line

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+                netdev_t *dev = msg.content.ptr;
+                dev->driver->isr(dev);
+            }
+                break;
```
here and at other places: please move break into braced scope

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+
+uint8_t semtech_loramac_join(uint8_t type)
+{
+    _semtech_loramac_call(_join, (void*)&type);
```
```(void *)&type```

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+    params.payload = tx_buf;
+    params.len = tx_len;
+
+    _semtech_loramac_call(_send, (void*)&params);
```
space befor star

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,631 @@
+    LoRaMacMibGetRequestConfirm(&mibReq);
+
+    if (!mibReq.Param.IsNetworkJoined) {
+        (void) rx_buf;
```
remove this line

dylad commented on pkg/semtech-loramac/contrib/semtech_loramac_radio.c:
```diff
@@ -0,0 +1,256 @@
+ *
+ * @note        Even if the function names are related to SX1276, they use
+ *              the RIOT SX127x driver and can be used with SX1272 and SX1276
+ *              devices.
```
I agree with that but what if someone want to reuse some old code using SX1272 ?
That person would have to rename everything ?

aabadie replied:
> what if someone want to reuse some old code using SX1272 ?

What do mean with "some old code using sx1272" ? In this PR using sx1276 or sx1272 is just a matter of using the corresponding driver module in the application `Makefile`, e.g `USEMODULE += sx1272` or `USEMODULE += sx1276`.
The upstream LoRaMAC code uses its own HAL and in the package we bypass it to reuse RIOT's internal driver. The RIOT driver avoids the code duplication that is done in LoRaMAC code.

dylad replied:
Sorry If I was unclear, what I mean is the following : 
Let's say someone what to reuse an old code using SX1272.
With the current state, functions like SX1272Init  cannot be used. 

I'm not saying it's a no-go (and I have no better proposition right now) but we should at least make it clear.

aabadie replied:
ok, I see. This use case is not covered unfortunately, and I think it won't be. In this PR, the proposal is to go through a custom RIOT API (see semtech-loramac.h) that provide basic functions for getting/setting mac parameters, joining a network and sending/receiving packets. This is more or less what is provided by LoRaMAC applications (see their main.c's)

dylad replied:
I guess it is fine for now. We should just write it somewhere to warn the current limitation of this pkg. What do you think ?

aabadie replied:
Personnally, I don't see it as a limitation but rather as an improvement regarding the original API provided by Semtech. Have you looked at their application code ? I just wrapped it in semtech-loramac.c (except the link check and compliance test) and added an API for mainpulating the MAC on top of their internal code. The MAC is isolated in a thread so that all calls to the MAC timers and other stuff are not processed in the ISR context (which is what they do basically) but in a separate thread context.
I think IMHO that in term of user experience, this is not comparable.

dylad replied:
> Have you looked at their application code

Unfortunately yes... I don't like it at all.

I guess the word 'limitation' was not a suitable one. 
People who will use this pkg  probability  have already used the original semtech loramac. This is why we must inform them (wiki page ?) the current differences (not limitations) from the original loramac.

aabadie replied:
> This is why we must inform them (wiki page ?) the current differences (not limitations) from the original loramac.

The simplest is to add note in the doc.txt of the package. This will be in the doxygen documentation.

aabadie replied:
I just pushed a detailed doxygen documentation with a warning.

dylad replied:
Thanks @aabadie !

jia200x replied:
ping @aabadie

dylad commented on pkg/semtech-loramac/contrib/semtech_loramac_radio.c:
```diff
@@ -0,0 +1,256 @@
+bool SX1276CheckRfFrequency(uint32_t frequency)
+{
+    (void) frequency;
+    /* Implement check. Currently all frequencies are supported */
```
TODO ?

aabadie replied:
This depends on the board, the upstream code [does the same](https://github.com/Lora-net/LoRaMac-node/blob/master/src/boards/SensorNode/sx1276-board.c#L216)

dylad commented on pkg/semtech-loramac/contrib/semtech_loramac_radio.c:
```diff
@@ -0,0 +1,256 @@
+    (void) freq;
+    (void) power;
+    (void) time;
+    /* TODO */
```
TODO ?

aabadie replied:
Yes, if you look at the upstream code ([sx1276](https://github.com/Lora-net/LoRaMac-node/blob/master/src/radio/sx1276/sx1276.c#L1124), [sx1272](https://github.com/Lora-net/LoRaMac-node/blob/master/src/radio/sx1272/sx1272.c#L972)), this is something that should be in the driver, and we don't support that for the moment. It also switch the radio modem to FSK mode and this mode is not supported by RIOT's driver. I don't know what are the implications.

dylad commented on tests/pkg_semtech-loramac/README.md:
```diff
@@ -0,0 +1,150 @@
+* The type of join procedure: ABP (Activation by personnalization) or OTAA (Over
+The Air Activation)
+* The device EUI: an 8 bytes array
+* The application EUI: an 8 bytes array
```
an or a ?

aabadie replied:
not sure, I think `an` is correct since 8 (eight) starts with a `e`

aabadie replied:
Seems like Google agrees btw.

dylad commented on tests/pkg_semtech-loramac/README.md:
````diff
@@ -0,0 +1,150 @@
+Depending on the type of radio device, set the `DRIVER` variable accordingly:
+For example:
+```
+    DRIVER=sx1272 make BOARD=nucleo-f411 -C pkg/semtech-loramac flash term
````
I tried this but it failed !

aabadie replied:
ok, I forgot to rename one internal variable, fixed.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+    semtech_loramac_set_netid(LORAMAC_DEFAULT_NETID);
+
+    mibReq.Type = MIB_DEV_ADDR;
+    mibReq.Param.DevAddr = ((uint32_t)dev_addr[0] << 24 |
```
memcpy? or possibly memcpy & hotns?

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+            _join_abp();
+            break;
+
+        default:
```
default case can be removed

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+
+static void _send(void *arg)
+{
+    loramac_send_params_t params = *(loramac_send_params_t*)arg;
```
space befor cast star

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+            break;
+
+        case NETDEV_EVENT_TX_COMPLETE:
+            sx127x_set_sleep((sx127x_t*)dev);
```
cast star

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+            size_t len;
+            len = dev->driver->recv(dev, NULL, 0, 0);
+            dev->driver->recv(dev, payload, len, &packet_info);
+            radio_events.RxDone((uint8_t*)payload, len, packet_info.rssi,
```
cast star. Last time I'll mention them for this file.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+    while (1) {
+        msg_t msg;
+        msg_receive(&msg);
+        switch(msg.type) {
```
space after switch

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+
+            default:
+                DEBUG("[semtech-loramac] Unexpected msg type '%04x'\n", msg.type);
+                break;
```
no break needed here

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+
+uint8_t semtech_loramac_join(uint8_t type)
+{
+    _semtech_loramac_call(_join, (void *)&type);
```
cast not needed

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+    params.payload = tx_buf;
+    params.len = tx_len;
+
+    _semtech_loramac_call(_send, (void *)&params);
```
cast not needed

kaspar030 commented on pkg/semtech-loramac/include/semtech_loramac.h:
```diff
@@ -0,0 +1,293 @@
+ * @brief   LoRaMAC status
+ */
+enum {
+    SEMTECH_LORAMAC_JOIN_SUCCEEDED = 0,          /**< Join procedure succeeded */
```
```= 0````not needed (first is implicitly 0)

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,628 @@
+                msg.content.value = SEMTECH_LORAMAC_TX_DONE;
+                msg_send(&msg, _semtech_loramac_handler_pid);
+            }
+                break;
```
break into scope please

kaspar030 commented:
Overall this looks quite OK.

@aabadie I'm a bit puzzled that I have to remark certain style remarks multiple times, you being our pyhon styler. E.g., if one pointer star needs changing, you should go through all the files to ensure the poor reviewers eyes don't have to see that again in the same PR! ;)

Some open questions:

- why can there be only one loramac state? no multi-radio possible?
- the global variables must be namespaced.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -67,32 +64,22 @@
-uint8_t nwk_skey[LORAMAC_NWKSKEY_LEN];
-uint8_t app_skey[LORAMAC_APPSKEY_LEN];
-uint8_t dev_addr[LORAMAC_DEVADDR_LEN];
+RadioEvents_t _semtech_loramac_radio_events;
```
Sorry if I wasn't clear. exported symbols starting with underscore are against C spec. use underscore for static, remove for exported global symbols.

aabadie commented:
> you should go through all the files to ensure the poor reviewers eyes don't have to see that again in the same PR!

sorry about that, lots of required changes, I just forgot a few. Next time I'll use uncrustify...

> why can there be only one loramac state? no multi-radio possible?

This is a limitation in the upstream code of the package itself. If we want the patches to stay miminal, we have to do that.

> the global variables must be namespaced

Should be done now.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac_arch.c:
```diff
@@ -0,0 +1,67 @@
+
+void srand1(uint32_t seed)
+{
+    random_init(seed);
```
can this be avoided? If ```srand1()``` is only called once at init time, maybe an empty stub is ok, considering that random_init() gets called by auto init.

aabadie replied:
You are right I think, I removed this call and still works.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac_arch.c:
```diff
@@ -0,0 +1,67 @@
+    }
+}
+
+int8_t Nibble2HexChar(uint8_t a)
```
is this still needed?

aabadie replied:
removed

aabadie commented:
About `pkg/semtech-loramac/Makefile.include` (I cannot comment in the above thread, weird)

> If added this way, all these includes will be applied globally (to all modules). Please confirm that this is necessary, or maybe can be shrunk a little.

This is the only place I found that works. Do you have an idea where it would be better ? (I already tested `pkg/semtech-loramac/Makefile` without success)

kaspar030 commented:
> This is the only place I found that works. Do you have an idea where it would be better ?

What does "works" mean? Find out which files are included where, and why the include path is needed.
Honestly I'd prefer symlinking the actually included headers somwhere instead of basically adding all loramac source folders to the global include path, but that should be a last resort.

aabadie commented:
@kaspar030, I found a way to only add the include path in the module directories. It builds correctly and is not added globally in `Makefile.include`.

aabadie commented:
> Where are memcpy1 and memset1 declared? Maybe that declaration can be patched into calling the proper functions.

They are declared in the loramac-node package build directory, in `semtech-loramac/src/boards/mcu/stm32/utilities.h`, we could patch `semtech-loramac/src/boards/mcu/stm32/utilities.c` with the implementation in `semtech-loramac-arch.c` but then the patches size will increase. I don't know what's the best solution.

kaspar030 commented:
> They are declared in the loramac-node package build directory, in semtech-loramac/src/boards/mcu/stm32/utilities.h, we could patch semtech-loramac/src/boards/mcu/stm32/utilities.c with the implementation in semtech-loramac-arch.c but then the patches size will increase. I don't know what's the best solution.

I think that would be the way to go: remove them from utilities.c, static inline them to the corresponding C functions in utilities.h.

kaspar030 commented:
> the way to go

Or, patch where used, which IMO would be optimal. We don't care about patch size, only about about patch maintainability.

kaspar030 commented:
> Or, patch where used, which IMO would be optimal.

scratch that, patching utilities.h is probably more maintainable.

aabadie commented:
@kaspar030, utilities.h/c are now patched in the package build directory and the _arch.c file is removed from RIOT code.

aabadie commented:
@kaspar030 @miri64 do you have remaining comments ? Can I squash ?

smlng commented:
looks like this is not WIP anymore, but rather in the final stages if fixing? Would be great to have it in the release.

jia200x commented:
@smlng @aabadie it's not linking with GCC 7:

```
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMac.o): In function `PrepareFrame':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMac.c:2068: undefined reference to `memcpyr'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMac.c:2070: undefined reference to `memcpyr'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMac.c:2123: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMac.c:2196: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMac.o): In function `OnRadioRxDone':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMac.c:1086: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMac.o): In function `LoRaMacMibSetRequestConfirm':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMac.c:2751: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMac.o): In function `LoRaMacMlmeRequest':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMac.c:3109: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMac.o): In function `LoRaMacMcpsRequest':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMac.c:3225: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMacCrypto.o): In function `LoRaMacPayloadEncrypt':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:114: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMacCrypto.o): In function `LoRaMacJoinDecrypt':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:173: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMacCrypto.o): In function `LoRaMacJoinComputeSKeys':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:188: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMacCrypto.o):/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:191: more undefined references to `memset1' follow
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_mac.a(LoRaMacCrypto.o): In function `LoRaMacJoinComputeSKeys':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:193: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:194: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:197: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:199: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/mac/LoRaMacCrypto.c:200: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_crypto.a(cmac.o): In function `AES_CMAC_Init':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:60: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:62: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_crypto.a(cmac.o): In function `AES_CMAC_Update':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:78: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:101: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:93: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:95: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/samr21-xpro/semtech_loramac_crypto.a(cmac.o): In function `AES_CMAC_Final':
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:110: undefined reference to `memset1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:148: undefined reference to `memcpy1'
/home/jialamos/Development/RIOT/tests/pkg_semtech-loramac/bin/pkg/samr21-xpro/semtech-loramac/src/system/crypto/cmac.c:150: undefined reference to `memset1'

```

kaspar030 commented on tests/pkg_semtech-loramac/main.c:
```diff
@@ -0,0 +1,427 @@
+ */
+
+/**
+ * @ingroup     test
```
s/test/tests/

kaspar030 commented on tests/pkg_semtech-loramac/main.c:
```diff
@@ -0,0 +1,427 @@
+            join_type = LORAMAC_JOIN_ABP;
+        }
+        else {
+            (void) join_type;
```
this line can be removed

kaspar030 commented on tests/pkg_semtech-loramac/main.c:
```diff
@@ -0,0 +1,427 @@
+                cnf = LORAMAC_TX_UNCNF;
+            }
+            else {
+                (void) cnf;
```
remove line

kaspar030 commented on tests/pkg_semtech-loramac/main.c:
```diff
@@ -0,0 +1,427 @@
+
+        semtech_loramac_rx_data_t rx_data;
+        switch (semtech_loramac_send(cnf, port,
+                                     (uint8_t*)argv[2], strlen(argv[2]),
```
here and below: pointer star space missing

kaspar030 commented on tests/pkg_semtech-loramac/README.md:
````diff
@@ -0,0 +1,150 @@
+```
+    > loramac tx This\ is\ RIOT! uncnf 10
+```
+* Switch the default datarate index (from 1 to 16). 5 is for SF7, BW125:
````
actual code and help allow *0* to 16.

kaspar030 commented on tests/pkg_semtech-loramac/Makefile:
```diff
@@ -0,0 +1,22 @@
+APPLICATION = pkg_semtech-loramac
```
application not necessary, will be set by Makefile.tests_common

kaspar030 commented on tests/pkg_semtech-loramac/Makefile:
```diff
@@ -0,0 +1,22 @@
+APPLICATION = pkg_semtech-loramac
+include ../Makefile.tests_common
+
+BOARD ?= b-l072z-lrwan1
```
This must go above including Makefile.tests_common, otherwise that file sets "native"

kaspar030 commented on pkg/semtech-loramac/patches/0003-adapt-utilities-functions.patch:
```diff
@@ -0,0 +1,162 @@
+  */
+-void memset1( uint8_t *dst, uint8_t value, uint16_t size );
++inline void memset1( uint8_t *dst, uint8_t value, uint16_t size )
++{
```
the idea was to call ```memset()``` here, not to make the code static inline

kaspar030 commented on pkg/semtech-loramac/patches/0003-adapt-utilities-functions.patch:
```diff
@@ -0,0 +1,162 @@
+-void memcpy1( uint8_t *dst, const uint8_t *src, uint16_t size );
++inline void memcpy1( uint8_t *dst, const uint8_t *src, uint16_t size )
++{
++    while (size--) {
```
call ```memcpy()``` here

aabadie commented:
@kaspar030, I just addressed your last comments. Regarding the utilities function patch, I removed the inline stuff and just patched the *.c file.
@jia200x I fixed the build issue with gcc 7.2 by adding another module for utilities functions.

I hope this is fine now.

smlng commented:
Murdock doesn't check GCC 7.2 right, hence I tested locally and can confirm that the linker issues is gone and build succeeds! Nice! 

@miri64 and @kaspar030 please act on your outstanding change requests today, they seem to be (sufficiently) addressed. @jia200x would be nice to get your ACK, too. 

IMHO this is ready and should go into the release 2018.01.

smlng commented:
@aabadie please squash after final approval by reviewers.

aabadie commented:
> please squash after final approval by reviewers.

I'll do that with pleasure :)

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,624 @@
+ * @file
+ * @brief       Implementation of public API for Semtech LoRaMAC
+ *
+ *              This implementation is an adaption of the applications provided
```
Seeing that you're not consistently following this indentation, please fix the long descriptions to align with the "@".

smlng replied:
I agree with @kaspar030, long/detail description must not align with `@brief`, but rather should align to one space after `*`, i.e., with the `@`s above and below for params.

kaspar030 commented on pkg/semtech-loramac/include/semtech-loramac/timer.h:
```diff
@@ -0,0 +1,129 @@
+void TimerReset(TimerEvent_t *obj);
+
+/**
+ * @brief   Set timer new timeout value
```
@briefs not consistently aligned. I propose three spaces after brief.

kaspar030 commented on pkg/semtech-loramac/include/semtech_loramac.h:
```diff
@@ -0,0 +1,303 @@
+ * @param[in] tx_len       The length of the TX buffer
+ * @param[out] rx_data     The RX data descriptor
+ *
+ * @return                 SEMTECH_LORAMAC_NOT_JOINED when the network is not joined
```
Here and everywhere else: lines are too long. I propose maybe changing the alignment after all.

smlng commented:
@kaspar030 I wasn't aware that we that strict with 

> fix line lengths to <=80

especially for documentation. Nevertheless, line length should not exceed it too much, but IMHO still readability first, but also `@brief`, `@params`, and `@returns` should be short anyway. If any of those needs longer description this should be done in the description below `@brief`.

aabadie commented:
doxygen updated

jia200x commented:
I will do a general review again before ACK

aabadie commented on pkg/semtech-loramac/patches/0003-adapt-utilities-functions.patch:
```diff
@@ -0,0 +1,162 @@
+  * \param [IN]  size Number of bytes to be copied
+  */
+-void memcpy1( uint8_t *dst, const uint8_t *src, uint16_t size );
++inline void memcpy1( uint8_t *dst, const uint8_t *src, uint16_t size )
```
the inline here is a copy paste issue from the previous patch version, I'll remove it.

aabadie replied:
Ok, you are reviewing an old version of the code, this is not in the patch anymore....

dylad commented on pkg/semtech-loramac/contrib/semtech_loramac_radio.c:
```diff
@@ -0,0 +1,256 @@
+/**
+ * LoRa function callbacks
+ */
+const struct Radio_s Radio =
```
Aaaaalright then I agree it would be better to rename with a generic name -> sx127x.
As I recall (correct me if I'm wrong), the Radio_s struct is define for SX1272 and SX1276 (In Semtech code) and as @jia200x said this is an abstraction layer. This level of abstraction exists because Semtech did not write generic functions from the beginning so function like SX1272Init and SX1276Init are indeed different right ? (because SX1272 and SX1276 are different chips)
In our case, we already wrote generic functions from the beginning so we only need the Radio_s struct to match Semtech application code. 
I would prefer to rename directly all theses functions to RIOT current API instead of keeping SX1276xxxx stuff as it's very confusing...

aabadie replied:
Just pushed the update

jia200x replied:
> As I recall (correct me if I'm wrong), the Radio_s struct is define for SX1272 and SX1276 (In Semtech code) and as @jia200x said this is an abstraction layer. This level of abstraction exists because Semtech did not write generic functions from the beginning so function like SX1272Init and SX1276Init are indeed different right ? (because SX1272 and SX1276 are different chips)

Exactly :)

aabadie commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,839 @@
+    dev->netdev.driver = &sx127x_driver;
+    dev->netdev.event_callback = _event_cb;
+
+    _semtech_loramac_handler_pid = thread_getpid();
```
Yes

aabadie replied:
But why your comments are appearing in old code ?

aabadie commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,839 @@
+    params.payload = tx_buf;
+    params.len = tx_len;
+
+    _semtech_loramac_call(_send, (void*)&params);
```
Why having a blocking call here ?

jia200x commented:
ACK

aabadie commented on tests/pkg_semtech-loramac/main.c:
```diff
@@ -0,0 +1,398 @@
+            puts("Usage: set public <on|off>");
+            return 1;
+        }
+        semtech_loramac_set_public_network(public);
```
I see the problem but changing that will takes some time. Normally the main thread is blocked when the MAC thread work (waiting for messages from the MAC when it's done).

aabadie commented:
@kaspar030 @miri64 shall I squash ?

kYc0o commented:
Just tested with a b-l072z-lrwan1 and works like a charm for common operation. I didn't do exhaustive tests for __all__ functionalities but we can send and receive LoRa packets without any issue.

You should just "rename" your first commit, you have a typo.

I also tried with the other sx127x modules and worked good so far on nucleo boards, however it failed on arduino-zero, but that's not a problem (anyways there's no SPI config for arduino-zero for the given pins).

I'd ACK it but I'll wait for the current reviewers comments and their ACK.

aabadie commented:
Thanks for testing @kYc0o. I fixed the typo in the first commit message.

smlng commented:
ping @kaspar030 and @miri64, please have another look here - this should go into the release.

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,624 @@
+
+        case NETDEV_EVENT_TX_TIMEOUT:
+            msg.type = MSG_TYPE_TX_TIMEOUT;
+
```
remove line here (or add one to NETDEV_EVENT_ISR)

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,624 @@
+
+        case NETDEV_EVENT_RX_TIMEOUT:
+            msg.type = MSG_TYPE_RX_TIMEOUT;
+
```
remove empty line

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,624 @@
+                netdev_t *dev = msg.content.ptr;
+                dev->driver->isr(dev);
+                break;
+            }
```
here and at other places: I think with a closing brace, the empty line after is not helping readability, so please remove them

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,624 @@
+#include <string.h>
+
+#include "msg.h"
+#include "shell.h"
```
shell isn't used in this file?

kaspar030 commented on pkg/semtech-loramac/contrib/semtech_loramac.c:
```diff
@@ -0,0 +1,624 @@
+#include "semtech-loramac/board.h"
+#include "LoRaMac.h"
+#include "region/Region.h"
+#include "utilities.h"
```
is that header includable with a path? ```utilities.h``` by itself sounds generic enough that it might lead to clashes otherwise.

aabadie commented:
I'll address the last comments @kaspar030

aabadie commented:
comments addressed @kaspar030, directly squashed and rebased. @miri64 is it ok for you ?

aabadie commented on pkg/semtech-loramac/Makefile.loramac_mac:
```diff
@@ -0,0 +1,5 @@
+MODULE := semtech_loramac_mac
+
+CFLAGS += -Wno-missing-field-initializers -Wno-unused-parameter -Wno-sign-compare
```
I just moved this CFLAGS rules here and [here](https://github.com/RIOT-OS/RIOT/pull/8264/files#diff-4f32d7da4e468a1350aea6570389f140R3) so that it's not applied globally to the package module

aabadie commented:
07aac76 is improving the CFLAGS use in the package modules

kaspar030 commented:
There's a typo in one of the commit messages (remane). Otherwise the code looks good now.

smlng commented:
> typo in one of the commit messages (remane)

@aabadie please fix on short notice, @kYc0o had pointed that out too, see above.

aabadie commented:
> There's a typo in one of the commit messages

Damn, I fixed it yesterday but on another personal computer and this morning I forced a push from my Inria laptop... overriding the change.
Btw, this is fixed now. @miri64 do you ACK?

kYc0o commented:
> no response in time to get this into release 2018.01.

Anyways, congrats @aabadie !

jia200x commented:
Congratz @aabadie!!

dylad commented:
Wonderful job @aabadie !

aabadie commented:
Wow !! This really makes my day :D
And many thanks to all people involved, mainly @jia200x and @fjmolinas for the initial work, but also reviewers!

fjmolinas commented:
Congrats @aabadie !!

miri64 commented:
Congrats @aabadie. But

> no response in time to get this into release 2018.01.

I had also had no time to check this... It leaves a bitter taste that my concerns about the architecture are just dismissed like that, just so that a (because of that potentially unstable) feature makes it into the release. As far as I can see my concerns are addressed, but I'm still not sure they actually are...

aabadie commented:
> It leaves a bitter taste that my concerns about the architecture are just dismissed like that

Don't worry, your concerns are WIP in another branch. I'll continue to work on that.

> a [...] potentially unstable feature makes it into the release

I understand your concerns but I would rather see it as a first step. Maybe we can just mention in the release notes that this is an experimental integration of LoRaWAN into RIOT ?
Also notice that the actual version of the package points to the latest release of Loramc-node and it'll have to be adapted to the coming one (that should be released soon).

miri64 commented:
> I understand your concerns but I would rather see it as a first step. Maybe we can just mention in the release notes that this is an experimental integration of LoRaWAN into RIOT ?

Agreed.
